Move serialization functions to include

PatKamin opened the pull request:
Related to issue #828 .

C API will be introduced in another PR.

pbalcer commented on scripts/templates/print.hpp.mako:
```diff
@@ -106,7 +108,9 @@ def findMemberType(_item):
-namespace ${x}_params {
+namespace ${x}_print {
+## API functions declarations #################################################
+// RFC: Should is_handle stay in the common dir?
```
as long as it's in a namespace, I don't see why not. Maybe we can move it to `details` (see below).

pbalcer commented on scripts/templates/print.hpp.mako:
```diff
@@ -117,49 +121,49 @@ template <> struct is_handle<${th.make_type_name(n, tags, obj)}> : std::true_typ
-template <typename T> inline void serializePtr(std::ostream &os, const T *ptr);
-template <typename T> inline void serializeFlag(std::ostream &os, uint32_t flag);
-template <typename T> inline void serializeTagged(std::ostream &os, const void *ptr, T value, size_t size);
+template <typename T> UR_DLLEXPORT inline ${x}_result_t UR_APICALL ${x}PtrPrint(std::ostream &os, const T *ptr);
```
Since this is a `print` header, my preference would be to name all these `urPrintSomething`. Makes it easier to find all print functions.

PatKamin replied:
I would prefer to call these functions this way, too. I just adhered to the function naming convention of UR API (https://github.com/oneapi-src/unified-runtime/blob/main/scripts/core/CONTRIB.rst#naming-convention). Can I ignore this convention in the case of a printing API then?

PatKamin replied:
Renamed

igchor replied:
But we now have something like `ur_print::urFunctionParamsPrint`, which seems a bit excessive. Why not just `ur::printFunctionParams()` (i.e. rename namespace to `ur`, remove `ur` prefix from the function name and possibly start the function name with `print`)

pbalcer replied:
`ur::print::functionParams()` or maybe `ur::extras::printFunctionParams()`?

I think I'd prefer this whole thing be in its own namespace. Hypothetically, we might want to add some more C++ functionality or bindings, and imho it should be possible to use that functionality without having to import all these operator overloads.

igchor replied:
Well, you don't have to include ur_print.hpp header if you don't want to. But I like the `ur::extras::printFunctionParams()` or perhaps (`ur::info::printFunctionParams()`)

pbalcer commented on scripts/templates/print.hpp.mako:
```diff
@@ -381,14 +389,14 @@ inline std::ostream &operator<<(std::ostream &os, [[maybe_unused]] const struct
-template <typename T> inline void serializePtr(std::ostream &os, const T *ptr) {
+///////////////////////////////////////////////////////////////////////////////
+// @brief Print pointer value
+template <typename T> inline ${x}_result_t ${x}PtrPrint(std::ostream &os, const T *ptr) {
```
This should not be part of the public API. Maybe we should have a `ur_print::details` namespace?

pbalcer commented on scripts/templates/print.hpp.mako:
```diff
@@ -401,9 +409,13 @@ template <typename T> inline void serializePtr(std::ostream &os, const T *ptr) {
-inline int serializeFunctionParams(std::ostream &os, uint32_t function, const void *params) {
+///////////////////////////////////////////////////////////////////////////////
+// @brief Print function parameters
+UR_DLLEXPORT inline ${x}_result_t UR_APICALL ${x}FunctionParamsPrint(std::ostream &os, uint32_t function, const void *params) {
```
I'm not sure if this is useful outside of urtrace (but I guess someone might want to implement a similar functionality themselves).

```suggestion
UR_DLLEXPORT inline ${x}_result_t UR_APICALL ${x}FunctionParamsPrint(std::ostream &os, enum ur_function_t function, const void *params) {
```

PatKamin replied:
Done

pbalcer commented on scripts/Doxyfile:
```diff
@@ -848,7 +848,8 @@ RECURSIVE              = YES
 
-EXCLUDE                = README.md
+EXCLUDE                = README.md \
+                         ../include/ur_print.hpp
```
Just occurred to me that these APIs won't have doxygen docs. We should eventually address this.

PatKamin replied:
Added generation of docs for all operators and a function from ::extras.

igchor commented on scripts/templates/print.hpp.mako:
```diff
@@ -117,49 +121,50 @@ template <> struct is_handle<${th.make_type_name(n, tags, obj)}> : std::true_typ
-template <typename T> inline void serializePtr(std::ostream &os, const T *ptr);
-template <typename T> inline void serializeFlag(std::ostream &os, uint32_t flag);
-template <typename T> inline void serializeTagged(std::ostream &os, const void *ptr, T value, size_t size);
+template <typename T> UR_DLLEXPORT inline ${x}_result_t UR_APICALL ${x}PrintPtr(std::ostream &os, const T *ptr);
```
Hm, do we actually need UR_DLLEXPORT? All those functions are defined in the header, right?

PatKamin commented:
> Hm, do we actually need UR_DLLEXPORT? All those functions are defined in the header, right? 

Replaced with `UR_APIEXPORT` for functions visibility on Windows. Also removed these macros from functions of the `::details` namespace.

PatKamin commented:
> Just occurred to me that these APIs won't have doxygen docs. We should eventually address this. 

Added generation of docs for all operators and a function from ::extras.

PatKamin commented:
> Well, you don't have to include ur_print.hpp header if you don't want to. But I like the `ur::extras::printFunctionParams()` or perhaps (`ur::info::printFunctionParams()`) 

Rearanged namespaces in the header.

kbenzie reviewed (approved):
Sorry I missed the review request last week. These changes are looking good 👍
